Feature parity with Fuzzel, Wofi, (maybe) Rofi. Proper Quoting and Reserved Characters Handling in Arguments

wqvi opened the pull request:
Fuzzel, Wofi, and (maybe) Rofi quotes in the Exec line. This patch allows for j4-dmenu-desktop to launch the Exec line with quotes. It is rough around the edges but it's a good start.

An example entry:
Exec="/usr/bin/foobar" "--baz"

j4-dmenu-desktop doesn't consume the second argument in the Exec line causing issues when trying to launch desktop files like that. From what I am aware of these kinds of desktop files are rare.

Brief summary from [The Exec key](https://specifications.freedesktop.org/desktop-entry-spec/1.5/exec-variables.html)
Double quotes should enclose arguments, and certain characters like quotes, backticks, dollar signs, and backslashes must be escaped with a backslash. Implementations need to undo this quoting before expanding field codes and passing arguments to programs. Reserved characters include space, tab, newline, various punctuation marks, and symbols like tilde and ampersand. Got it!

meator commented:
Isn't this feature already implemented & functional?

The special `Exec` key handling is not handled in `Application.hh`. The `Application` class treats the `Exec` key like any other key, so it only escapes `\\`, `\t`, `\s` etc. (as specified in the standard).

The `Executable` classes (descendants of the `BaseExecutalbe` class) defined in `main.cc`[^1] take the unprocessed `Exec` key of `Application`[^2] and then "assemble" it into a suitable `argv` format, which is then `exec()`ed.

The processing of `Exec` is mainly done in `CMDLineAssembler.cc` in [`CMDLineAssembly::convert_exec_to_command()`](https://github.com/enkore/j4-dmenu-desktop/blob/595dd14660697758f3c4c94958128b9fe311443c/src/CMDLineAssembler.cc#L67).

[^1]: this could be put somewhere else, `main.cc` contains too much logic in it as is
[^2]: only basic escapes are handled as mentioned above

wqvi commented:
Yes, this is partially true, but it is missing feature parity with fuzzel and wofi, and (presumably) rofi. This commit includes the handling of single quotes, which I believe is out of spec. Fuzzel and wofi definitely handle single quotes despite it being out of spec. I've changed the code to be in a more appropriate location.

wqvi committed:
Added handling of single quotes.

This is for feature parity with fuzzel and wofi. It is technically out
of spec.

meator commented:
Your pull request description is highly ambiguous and unclear.

As I've explained, j4-dmenu-desktop handled double quotes correctly in accordance to freedesktop's Desktop Entry Specification, which you have linked.

Quoting using single quotes (`'`), which appears to be the focus of this pull request (even though it doesn't clearly state this intent), is not defined in the Desktop Entry Specification. Endorsing such behavior will open up a can of worms I don't really want to touch.

The reason why the behavior of j4-dmenu-desktop differs from the behavior of other launchers is most likely the handling of desktop file execution. I have held lengthy discussions about this behavior on [the XDG mailing list](https://lists.freedesktop.org/archives/xdg/2024-June/014718.html)[^1]. The behavior difference is caused by the ambiguity of the standard. I have complained about this on the mailing list and a solution was proposed, but I don't think this will be addressed anytime soon.

I will **not** endorse nonstandard behavior just because other desktop file launchers do it. I am willing to consider doing that if there are popular desktop files which rely on this behavior or if recent changes to j4-dmenu-desktop broke some desktop files for some users (my commitment to this goal is described in #181).

If other desktop file launchers choose to implement nonconformant behavior, so be it, but it is undermining the authority of the standard. If enough implementations do it, the standard will become useless. I do not want to contribute to this.

I will close this pull request for the reasons described above. I am open to discussion and willing to reopen this pull request if there is a good reason to implement this change (providing examples of desktop files which rely on single quote quoting would help).

[^1]: Mainly the later replies to this thread. Also see [this thread](https://lists.freedesktop.org/archives/xdg/2022-August/014620.html) if you're interested.
